### PR TITLE
Fix finding controllers

### PR DESCRIPTION
**Pull request by alexdavid**
This is blocking refactoring testing on configuration service since putting spec files in controllers dir causes exprestive to throw an error

@kevgo @charlierudolph @jondistad please review when you get a chance.

Thanks

### CODE REVIEW

File: test/feature/support/helpers.coffee [hunk 1]
```diff
@@ -61,8 +61,15 @@ class Helpers
     child.on 'message', (message) ->
       done() if message is 'server started'
 
+    exitSuccessful = no
```
Comment (charlierudolph) [inline test/feature/support/helpers.coffee]: How about `unexpectedExit` with a default of yes and set it to no when you kill it?

File: test/feature/support/helpers.coffee [hunk 2]
```diff
@@ -61,8 +61,15 @@ class Helpers
...
     # After test are done kill the child
-    @cleanUpActions.push -> child.kill()
+    @cleanUpActions.push ->
+      exitSuccessful = yes
+      child.kill()
+
+    # Throw an error if we didn't exit from the cleanUpAction
+    child.on 'close', (err, signal) ->
```
Comment (charlierudolph) [inline test/feature/support/helpers.coffee:71]: Why not just use `child.on 'error'` instead using your own variable?

**Comment by charlierudolph** [comment]
two minor comments. :ship: when ready

File: src/exprestive.coffee [hunk 3]
```diff
@@ -47,6 +47,23 @@ class Exprestive
       @controllers[camelCase controllerName][controllerAction] arguments...
 
 
+  # Returns if the passed filePath resembles a controller
```
Comment (kevgo) [inline src/exprestive.coffee]: Replace `if` with `whether`, and add an empty line after the first line of the comment.

File: src/exprestive.coffee [hunk 4]
```diff
@@ -47,6 +47,23 @@ class Exprestive
...
+  # Controllers are .js or .coffee files that export a constructor and name
+  fileIsController: (filePath) ->
+    # Non-JS files aren't controllers
+    extension = path.extname filePath
+    return false unless extension is '.coffee' or extension is '.js'
```
Comment (kevgo) [inline src/exprestive.coffee]: replace `false` with `no`

File: src/exprestive.coffee [hunk 5]
```diff
@@ -47,6 +47,23 @@ class Exprestive
...
+
+    # Controllers must export a name
+    maybeController = require filePath
+    return false unless maybeController.name?
+
+    # Controllers must export a constructor function
+    return false unless typeof maybeController is 'function'
+
+    true
```
Comment (kevgo) [inline src/exprestive.coffee]: replace all boolean values with yes/no notation

File: src/exprestive.coffee [hunk 6]
```diff
@@ -47,6 +47,23 @@ class Exprestive
...
+    # Non-JS files aren't controllers
+    extension = path.extname filePath
+    return false unless extension is '.coffee' or extension is '.js'
+
+    # Controllers must export a name
+    maybeController = require filePath
+    return false unless maybeController.name?
+
+    # Controllers must export a constructor function
+    return false unless typeof maybeController is 'function'
```
Comment (kevgo) [inline src/exprestive.coffee]: Some other remarkable things about controllers is that they export a class whose name ends with `Controller`. This can be checked through `maybeController.name`

File: src/exprestive.coffee [hunk 7]
```diff
@@ -47,6 +47,23 @@ class Exprestive
       @controllers[camelCase controllerName][controllerAction] arguments...
 
 
+  # Returns if the passed filePath resembles a controller
+  # Controllers are .js or .coffee files that export a constructor and name
+  fileIsController: (filePath) ->
```
Comment (kevgo) [inline src/exprestive.coffee]: This could be extracted into a dedicated class, so that you can test this. There are tons of edge cases to be sure about here.

``` coffeescript
# Recognizes whether a given file is an Exprestive controller
class ControllerDetector

  constructor: (filePath) ->

  # Returns whether the file is an Expressive controller
  isController: ->
    isJsFile() and exportsName() and exportsConstructorFunction() and nameEndsWithController()

  isJsFile: ->
    extension = path.extname filePath
    extension is '.coffee' or extension is '.js'

  exportsName: ->
    maybeController = require filePath
    maybeController.name?

  exportsConstructorFunction: ->
    maybeController = require filePath
    typeof maybeController is 'function'

  nameEndsWithController: ->
    ...
```

File: src/exprestive.coffee [hunk 8]
```diff
@@ -75,8 +92,10 @@ class Exprestive
     @controllers = @options.controllers ? {}
     return if @options.controllers?
 
-    for file in fs.readdirSync @controllersDirPath
-      Controller = require path.join @controllersDirPath, file
+    for fileName in fs.readdirSync @controllersDirPath
```
Comment (kevgo) [inline src/exprestive.coffee]: Use async `fs.readdir` here

File: test/feature/finding_controllers.feature [hunk 9]
```diff
@@ -36,6 +36,25 @@ Feature: Finding controllers
...
+      """
+    And an exprestive app using defaults
+    Then the app doesn't error
+
+
+  Scenario: an application with a non-javascript file in controllers
+    Given a file "controllers/readme.md" with the content
+      """
+      Readme about controllers
+      =======================
```
Comment (kevgo) [inline test/feature/finding_controllers.feature]: Remove this line

File: test/feature/step_definitions/setup_definitions.coffee [hunk 10]
```diff
@@ -34,3 +34,7 @@ module.exports = ->
     @createFile "routes.coffee", routesFileContents, (err) =>
       return done.fail err if err
       done()
+
+
+  @Then /^the app doesn't error$/, (done) ->
```
Comment (kevgo) [inline test/feature/step_definitions/setup_definitions.coffee:39]: Simplify to `@Then /^the app doesn't error$/, done`

Comment (kevgo) [inline test/feature/step_definitions/setup_definitions.coffee:39]: Or add some verification here

File: test/feature/support/helpers.coffee (hunk 1 above)
Comment (alexdavid) [inline test/feature/support/helpers.coffee]: I like that :+1:

File: test/feature/support/helpers.coffee (hunk 2 above)
Comment (alexdavid) [inline test/feature/support/helpers.coffee:71]: I tried that, it didn't seem to be triggering the error event. I'll look more into this

File: test/feature/step_definitions/setup_definitions.coffee (hunk 10 above)
Comment (alexdavid) [inline test/feature/step_definitions/setup_definitions.coffee:39]: where is `done` defined?

File: src/exprestive.coffee (hunk 6 above)
Comment (alexdavid) [inline src/exprestive.coffee]: Do we want to silently skip controllers who's name doesn't end with `Controller`?

File: src/exprestive.coffee (hunk 8 above)
Comment (alexdavid) [inline src/exprestive.coffee]: will do in a future PR

File: test/feature/support/helpers.coffee (hunk 1 above)
Comment (alexdavid) [inline test/feature/support/helpers.coffee]: done :checkered_flag:

File: src/exprestive.coffee (hunk 3 above)
Comment (alexdavid) [inline src/exprestive.coffee]: done :checkered_flag:

File: src/exprestive.coffee (hunk 4 above)
Comment (alexdavid) [inline src/exprestive.coffee]: done :checkered_flag:

File: src/exprestive.coffee (hunk 5 above)
Comment (alexdavid) [inline src/exprestive.coffee]: done :checkered_flag:

File: test/feature/finding_controllers.feature (hunk 9 above)
Comment (alexdavid) [inline test/feature/finding_controllers.feature]: done :checkered_flag:

**Comment by alexdavid** [comment]
@kevgo @charlierudolph please take another look when you have a chance. Thanks!

File: src/controller_store.coffee [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+
+class ControllerStore
+
+  constructor: (@controllersDir, @dependencies) ->
+    @controllers = {}
+
+
+  # Calls a controller action with args
+  applyAction: ({controller, action, args}) ->
```
Comment (charlierudolph) [inline src/controller_store.coffee]: I think this should be `controllerName` instead of `controller`

File: src/exprestive.coffee [hunk 12]
```diff
@@ -44,7 +47,10 @@ class Exprestive
   # Registers a route on @middlewareRouter
   addRoute: ({httpMethod, url, controllerName, controllerAction}) ->
     @middlewareRouter[httpMethod.toLowerCase()] url, =>
-      @controllers[camelCase controllerName][controllerAction] arguments...
+      @controllers.applyAction
+        controller: controllerName
+        action: controllerAction
+        args: arguments
```
Comment (charlierudolph) [inline src/exprestive.coffee]: Not important but could possible make this one line with a couple renames

``` coffee
@middlewareRouter[httpMethod.toLowerCase()] url, (args...) =>
   @controllers.applyAction {controllerName, controllerActions, args}
```

**Comment by charlierudolph** [comment]
Looks good

File: src/controller_store.coffee (hunk 11 above)
Comment (alexdavid) [inline src/controller_store.coffee]: Done :checkered_flag:

File: src/exprestive.coffee (hunk 12 above)
Comment (alexdavid) [inline src/exprestive.coffee]: Done :checkered_flag:

**Comment by alexdavid** [comment]
Blocking on this change so I'm going to go ahead and ship this and release `0.5.0`. If there are any additional changes I'll release them as patches.

File: src/exprestive.coffee (hunk 6 above)
Comment (kevgo) [inline src/exprestive.coffee]: Maybe show a warning?

File: src/file_identifier.coffee [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+  constructor: (@filePath) ->
+
+
+  exportsConstructor: ->
+    typeof require(@filePath) is 'function'
+
+
+  exportsName: ->
+    {name} = require @filePath
+    typeof name is 'string'
```
Comment (kevgo) [inline src/file_identifier.coffee:15]: You can inline the temp variable here.

``` coffeescript
typeof require(@filePath).name is 'string'
```

File: src/controller_store.coffee [hunk 14]
```diff
@@ -0,0 +1,28 @@
+camelCase = require 'camel-case'
+FileIdentifier = require './file_identifier'
+fs = require 'fs'
+path = require 'path'
+
+
+class ControllerStore
```
Comment (kevgo) [inline src/controller_store.coffee:7]: I would add a description to this class. The purpose of a `ControllerStore` is not immediately obvious to the uninitiated reader.

File: src/controller_store.coffee [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+
+  # Calls a controller action with args
+  applyAction: ({controllerName, actionName, args}) ->
+    @controllers[camelCase controllerName][actionName] args...
+
+
+  # Populates @controllers by instantiating controllers found in @controllersDir
+  initialize: ->
+    for fileName in fs.readdirSync @controllersDir
```
Comment (kevgo) [inline src/controller_store.coffee:20]: Use async file operation. Exprestive is getting included into the middle of a larger code base, and you don't want to be the guy who blocks the whole event queue.

File: src/file_identifier.coffee [hunk 16]
```diff
@@ -0,0 +1,28 @@
+path = require 'path'
+
+
+class FileIdentifier
```
Comment (kevgo) [inline src/file_identifier.coffee:4]: Add a comment

File: test/feature/finding_controllers.feature [hunk 17]
```diff
@@ -34,3 +34,21 @@ Feature: Finding controllers
...
+    Then the app doesn't error
+
+
+  Scenario: an application with a non-javascript file in controllers
+    Given a file "controllers/readme.md" with the content
+      """
+      Readme about controllers
+      """
+    And an exprestive app using defaults
+    Then the app doesn't error
```
Comment (kevgo) [inline test/feature/finding_controllers.feature:54]: You have 3 criteria for controllers (@isJsFile() and @exportsName() and @exportsConstructor()), but only 2 scenarios here.

File: test/feature/step_definitions/setup_definitions.coffee (hunk 10 above)
Comment (kevgo) [inline test/feature/step_definitions/setup_definitions.coffee:39]: Haha, okay `done` doesn't exist here. But the empty body is weird here. Doesn't this step always pass, whether there are errors or not?

File: src/controller_store.coffee (hunk 15 above)
Comment (alexdavid) [inline src/controller_store.coffee:20]: Yeah, this should be pulled into a separate PR since it's a bigger change. We'll want to block http requests until all controllers are initialized.

Also for most applications using exprestive this will run once as the application boots up so it shouldn't really be blocking anything important anyway

But I do agree that we should eventually do this

File: test/feature/step_definitions/setup_definitions.coffee (hunk 10 above)
Comment (alexdavid) [inline test/feature/step_definitions/setup_definitions.coffee:39]: Yeah it doesn't actually do anything, but does it read OK to have a scenario with only 2 Givens and no Thens?

``` cucumber
Given a file "controllers/my_controller_spec.coffee" with the content
     """
     # This file doesn't export a controller
     """
And an exprestive app using defaults
```

vs

``` cucumber
Given a file "controllers/my_controller_spec.coffee" with the content
     """
     # This file doesn't export a controller
     """
And an exprestive app using defaults
Then the app doesn't error
```

Comment (alexdavid) [inline test/feature/step_definitions/setup_definitions.coffee:39]: I think when I do my refactor of lazy boot up then this step definition will cease to be empty

File: src/controller_store.coffee (hunk 15 above)
Comment (kevgo) [inline src/controller_store.coffee:20]: The thing it blocks is all other parts of the app initializing and booting up, thereby leading to increased startup times of any server that uses exprestive. That's bad imho.

Looking forward to seeing how you'll block HTTP requests!

File: test/feature/step_definitions/setup_definitions.coffee (hunk 10 above)
Comment (kevgo) [inline test/feature/step_definitions/setup_definitions.coffee:39]:
``` cucumber
Given a file "controllers/my_controller_spec.coffee" with the content
  """
  # This file doesn't export a controller
  """
Then an exprestive app using defaults starts up without error
```
